osu.Game/Overlays/Mods/ModSelectOverlay.cs

```diff
@@ -111,6 +111,10 @@ protected virtual IEnumerable<ShearedButton> CreateFooterButtons()
 
         private readonly Bindable<Dictionary<ModType, IReadOnlyList<Mod>>> globalAvailableMods = new Bindable<Dictionary<ModType, IReadOnlyList<Mod>>>();
 
+        public IEnumerable<Mod> AllAvailableAndValidMods => allAvailableMods
+                                                            .Select(s => s.Mod)
+                                                            .Where(m => isValidMod(m));
```

Comment:
I don't really like that this is exposed. Feels much too specific.

I'd propose the following equivalent instead:

```diff
diff --git a/osu.Game/Overlays/Mods/ModSelectOverlay.cs b/osu.Game/Overlays/Mods/ModSelectOverlay.cs
index ba3f01a688..7ec108e3ec 100644
--- a/osu.Game/Overlays/Mods/ModSelectOverlay.cs
+++ b/osu.Game/Overlays/Mods/ModSelectOverlay.cs
@@ -111,11 +111,7 @@ protected virtual IEnumerable<ShearedButton> CreateFooterButtons()
 
         private readonly Bindable<Dictionary<ModType, IReadOnlyList<Mod>>> globalAvailableMods = new Bindable<Dictionary<ModType, IReadOnlyList<Mod>>>();
 
-        public IEnumerable<Mod> AllAvailableAndValidMods => allAvailableMods
-                                                            .Select(s => s.Mod)
-                                                            .Where(m => isValidMod(m));
-
-        private IEnumerable<ModState> allAvailableMods => AvailableMods.Value.SelectMany(pair => pair.Value);
+        public IEnumerable<ModState> AllAvailableMods => AvailableMods.Value.SelectMany(pair => pair.Value);
 
         private readonly BindableBool customisationVisible = new BindableBool();
 
@@ -386,7 +382,7 @@ private void createLocalMods()
 
         private void filterMods()
         {
-            foreach (var modState in allAvailableMods)
+            foreach (var modState in AllAvailableMods)
                 modState.ValidForSelection.Value = modState.Mod.HasImplementation && IsValidMod.Invoke(modState.Mod);
         }
 
@@ -411,7 +407,7 @@ private void updateCustomisation()
             bool anyCustomisableModActive = false;
             bool anyModPendingConfiguration = false;
 
-            foreach (var modState in allAvailableMods)
+            foreach (var modState in AllAvailableMods)
             {
                 anyCustomisableModActive |= modState.Active.Value && modState.Mod.GetSettingsSourceProperties().Any();
                 anyModPendingConfiguration |= modState.PendingConfiguration;
@@ -468,7 +464,7 @@ private void updateFromExternalSelection()
 
             var newSelection = new List<Mod>();
 
-            foreach (var modState in allAvailableMods)
+            foreach (var modState in AllAvailableMods)
             {
                 var matchingSelectedMod = SelectedMods.Value.SingleOrDefault(selected => selected.GetType() == modState.Mod.GetType());
 
@@ -495,7 +491,7 @@ private void updateFromInternalSelection()
             if (externalSelectionUpdateInProgress)
                 return;
 
-            var candidateSelection = allAvailableMods.Where(modState => modState.Active.Value)
+            var candidateSelection = AllAvailableMods.Where(modState => modState.Active.Value)
                                                      .Select(modState => modState.Mod)
                                                      .ToArray();
 
diff --git a/osu.Game/Screens/OnlinePlay/FooterButtonFreeMods.cs b/osu.Game/Screens/OnlinePlay/FooterButtonFreeMods.cs
index 56a69be741..294c80677d 100644
--- a/osu.Game/Screens/OnlinePlay/FooterButtonFreeMods.cs
+++ b/osu.Game/Screens/OnlinePlay/FooterButtonFreeMods.cs
@@ -97,7 +97,7 @@ protected override void LoadComplete()
         /// </summary>
         private void toggleAllFreeMods()
         {
-            var availableMods = freeModSelectOverlay.AllAvailableAndValidMods.ToArray();
+            var availableMods = allAvailableAndValidMods.ToArray();
 
             Current.Value = Current.Value.Count == availableMods.Length
                 ? Array.Empty<Mod>()
@@ -108,7 +108,7 @@ private void updateModDisplay()
         {
             int current = Current.Value.Count;
 
-            if (current == freeModSelectOverlay.AllAvailableAndValidMods.Count())
+            if (current == allAvailableAndValidMods.Count())
             {
                 count.Text = "all";
                 circle.FadeColour(colours.Yellow, 200, Easing.OutQuint);
@@ -124,5 +124,9 @@ private void updateModDisplay()
                 circle.FadeColour(colours.Gray4, 200, Easing.OutQuint);
             }
         }
+
+        private IEnumerable<Mod> allAvailableAndValidMods => freeModSelectOverlay.AllAvailableMods
+                                                                                 .Where(state => state.ValidForSelection.Value)
+                                                                                 .Select(state => state.Mod);
     }
 }

```

Reply:
Aha, I didn't realise that the validity was included in `ModState`. I tried this at first but realised it was selecting invalid mods.